state-sync/storage-service/server/src/subscription.rs

```diff
@@ -0,0 +1,926 @@
+        let highest_known_version = subscription_request.highest_known_version();
+        let highest_known_epoch = subscription_request.highest_known_epoch();
+        let subscription_stream_id = subscription_request.subscription_stream_id();
+
+        // Create a new set of pending subscription requests using the first request
+        let mut pending_subscription_requests = BTreeMap::new();
+        pending_subscription_requests.insert(
+            subscription_request.subscription_stream_index(),
+            subscription_request,
+        );
+
+        Self {
+            highest_known_version,
+            highest_known_epoch,
+            next_index_to_serve: 0,
+            pending_subscription_requests,
+            subscription_stream_id,
+            last_stream_update_time: time_service.now(),
+            time_service,
+        }
+    }
+
+    /// Adds a subscription request to the existing stream. If this operation
+    /// fails, the request is returned to the caller so that the client
+    /// can be notified of the error.
+    pub fn add_subscription_request(
+        &mut self,
+        storage_service_config: StorageServiceConfig,
+        subscription_request: SubscriptionRequest,
+    ) -> Result<(), (Error, SubscriptionRequest)> {
+        // Verify that the subscription request stream ID is valid
+        let subscription_stream_id = subscription_request.subscription_stream_id();
+        if subscription_stream_id != self.subscription_stream_id {
+            return Err((
+                Error::InvalidRequest(format!(
+                    "The subscription request stream ID is invalid! Expected: {:?}, found: {:?}",
+                    self.subscription_stream_id, subscription_stream_id
+                )),
+                subscription_request,
+            ));
+        }
+
+        // Verify that the subscription request index is valid
+        let subscription_request_index = subscription_request.subscription_stream_index();
+        if subscription_request_index < self.next_index_to_serve {
+            return Err((
+                Error::InvalidRequest(format!(
+                    "The subscription request index is too low! Next index to serve: {:?}, found: {:?}",
+                    self.next_index_to_serve, subscription_request_index
+                )),
+                subscription_request,
+            ));
+        }
+
+        // Verify that the request known version and epoch are valid
+        let request_known_version = subscription_request.highest_known_version();
+        let request_known_epoch = subscription_request.highest_known_epoch();
+        if let Some(first_pending_request) = self.first_pending_request() {
+            let pending_known_version = first_pending_request.highest_known_version();
+            let pending_known_epoch = first_pending_request.highest_known_epoch();
+
+            // Verify the new and pending requests have the same known version and epoch
```

Comment:
Discussed offline, but the summary is:

The `pending_known_version` and `pending_known_epoch` on each request are irrelevant, except for the first request (i.e., when the stream is created). This is what the server uses to track the client's progress (i.e., each time the server sends a new data chunk). The reason for the invariant is to make it obvious that the client cannot and should not update the known versions during a stream, because, the server is tracking things, and there may be multiple responses currently in flight that the client hasn't processed. Thus, any updates from the client side may be wrong.

There's two reasons this is beneficial: (i) technically, it's possible that we receive client requests out-of-order, e.g., we get the 2nd request before the 1st, and in this case the logic for creating the stream is easy (we just use whatever we get first as the starting point); and (ii) if we make it optional, then we need to enforce that the first request always contains numbers, and the rest don't (which is a similar amount of code).

Reply:
This makes sense. I'd actually want to make this more obvious. The confusion for me is because it's sent on every request, but it's actually data for the entire stream. Two possible options
1. Rename the variable to `known_version_at_stream_creation`, `subscription_stream_known_version` or something prettier but still descriptive
2. Combine this with the `subscription_stream_id: u64` into a struct such as `StreamMetadata`

Reply:
Sounds, good. I've done both 😄 i.e., `SubscriptionStreamMetadata` now contains those two (renamed) variables as well as the stream ID. Let me know if you think this is clearer.

Reply:
Do you think it makes sense for `SubscriptionStreamRequests` to also contain a `SubscriptionStreamMetadata` instead of the raw `subscription_stream_id`? Then this check could be done on `self.subscrioption_stream_metadata` itself.

I feel it a bit awkward to verify on the `first_pending_request`. There's a corner case where the `first_pending_request` doesn't exist, so this check is not done -- if this is the case the values could change unnoticed.